projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,363 @@
+        assert isinstance(node_ids, list)
+        for val in node_ids:
+            assert val & 0xff == val
+            assert val < 64
+        self._node_ids = node_ids
+
+    def get_data(self):
+        """This function retrieves the data."""
+        return self._data
+
+    def set_data(self, data):
+        """This function sets the data."""
+        assert isinstance(data, bytearray)
+        self._data = data
+
+    def _check_kwargs(self, **kwargs):
+        """ This function checks that all variables are as expected"""
+
+        args = [
+            "protocol_version",
+            "datagram_type_id",
+            "node_ids",
+            "data"]
+        values = ["protocol_version", "datagram_type_id"]
+
+        # Check all arguments are present
+        for arg in args:
+            assert arg in kwargs
+
+        # Check that types are as expected
+        for value in values:
+            assert not isinstance(kwargs[value], list)
+        assert isinstance(kwargs["node_ids"], list)
+        assert isinstance(kwargs["data"], bytearray)
+
+        # Verify all inputs
+        assert kwargs["protocol_version"] & 0xff == kwargs["protocol_version"]
+        assert kwargs["datagram_type_id"] & 0xff == kwargs["datagram_type_id"]
+
+    def _convert_to_bytearray(self, in_value, bytes):
+        out_bytearray = bytearray()
+        for i in range(0, bytes):
+            out_bytearray.append((in_value >> (8 * i)) & 0xff)
+        return out_bytearray
+
+    def _convert_from_bytearray(self, in_bytearray, bytes):
+        value = 0
+        for i in range(0, bytes):
+            value = value | ((in_bytearray[i] & 0xff) << (i * 8))
+        return value
+
+
+class DatagramSender:
+    """This class acts as a distributor for the Datagram class on a bus."""
+
+    def __init__(self, bustype="socketcan", channel=DEFAULT_CHANNEL,
+                 bitrate=CAN_BITRATE, receive_own_messages=False):
+        print("Initializing CAN Bus...")
+        self.bus = can.interface.Bus(
+            bustype=bustype,
+            channel=channel,
+            bitrate=bitrate,
+            receive_own_messages=receive_own_messages)
+
+    def send(self, message):
+        """This sends the Datagrams."""
+
+        assert isinstance(message, Datagram)
+
+        chunk_messages = self._chunkify(message.serialize(), 8)
+        can_messages = []
+        message_arbitration_id = CAN_START_ARBITRATION_ID
+        message_extended_arbitration = False
+
+        # Populate an array with the can message from the library
+        for chunk_message in chunk_messages:
+            can_messages.append(can.Message(arbitration_id=message_arbitration_id,
+                                            data=chunk_message,
+                                            is_extended_id=message_extended_arbitration))
+
+            # After the first message, set the arbitration ID to 0
+            message_arbitration_id = CAN_ARBITRATION_ID
```

Comment:
One more thing - I think that Mitchell's C implementation assumes that the start message has no data, so it would probably be good to send an empty message with `CAN_START_ARBITRATION_ID` first then send all the messages with data with `CAN_ARBITRATION_ID`.

Reply:
What is meant by no data? Does that mean no node id's as well? Is it just a CAN message where the data is all 0's?

Reply:
It's a CAN message where there is no data, the DLC (data-length-code) is 0. (Nothing to do with the node IDs or anything with the structure of the datagram protocol, just a CAN message with no data. Like send an empty array instead of a list of 8 bytes of data.)